src/api/clock.rs

```diff
@@ -13,21 +13,19 @@
 // limitations under the License.
 
 pub trait Clock: Sized {
-    type Timer;
+    /// Stores data for the clock to recognize if this timer is elapsed or not.
+    ///
+    /// The Clock does not keep track of the timers it creates. Therefore, they should not wrap
+    /// unexpectedly. A timer that is elapsed may never return to a non-elapsed state.
+    ///
+    /// A default Timer should return `true` when checked with `is_elapsed`.
+    type Timer: Default;
 
     /// Creates a new timer that expires after the given time in ms.
     fn make_timer(&mut self, milliseconds: usize) -> Self::Timer;
 
-    /// Consumes a timer, and returns it if not expired.
-    fn check_timer(&mut self, timer: Self::Timer) -> Option<Self::Timer>;
-
-    /// Checks the timer and takes it, if expired.
-    fn update_timer(&mut self, timer: &mut Option<Self::Timer>) {
-        *timer = timer.take().and_then(|t| Self::check_timer(self, t));
-    }
-
-    /// Tickles the clock and makes it check its timers. Often a NOOP.
-    fn tickle(&mut self);
+    /// Checks whether a given timer is expired.
+    fn is_elapsed(&mut self, timer: &Self::Timer) -> bool;
```

Comment:
nit: It's pedantic but I think it's useful to precise the behavior with the following:

```rust
/// Until a timer expires, this function consistently returns false. Once it expires, this function consistently
/// returns true. In particular, it is valid to continue calling this function after the first time it returns true.
```

This is for coherence with the `type Timer` documentation.

---

For more explanation (using contrast), an alternative documentation (which would permit the `Clock` to not need to track absolute time) would be:

```rust
/// Represents a running timer.
///
/// Timers can only be created with `make_timer` because the `Clock` may keep track of the running timers.
///
/// A timer is running until `is_elapsed` returns true. After this, the timer should not be used again.
type Timer;

/// Returns whether a running timer has expired.
///
/// This function consistently returns false until the timer expires. After which it will return true. Once it
/// returns true, the timer should be dropped. In particular, this function should not be called again on
/// that timer.
fn is_elapsed(&mut self, timer: &Self::Timer) -> bool;
```

But I don't think this is needed because both `TockEnv` and `TestEnv` are implemented using absolute time. This is just to explain why additional documentation on `is_elapsed` is useful.

Reply:
Done